### PR TITLE
Go: Make flow configurations use new data flow API

### PR DESCRIPTION
**Pull request by owen-mc**
I highly recommend reviewing commit by commit (and spreading it out over several sessions). All configurations have tests (normally of the related query). There are no changes in results, though sometimes there are changes in exact `PathNode`s and edges between them.

### CODE REVIEW

**Comment by owen-mc** [comment]
I will rebase on main when https://github.com/github/codeql/pull/13822 is merged, which will fix the sync files check.

File: go/ql/lib/semmle/go/security/IncorrectIntegerConversionLib.qll [hunk 1]
```diff
@@ -148,6 +150,115 @@ class ConversionWithoutBoundsCheckConfig extends TaintTracking::Configuration {
   }
 }
 
+/** Flow state for ConversionWithoutBoundsCheckConfig. */
+newtype MyFlowState =
```
Comment (mbg) [inline go/ql/lib/semmle/go/security/IncorrectIntegerConversionLib.qll:154]: Should this have a less generic name, e.g. `IntegerConversionState`?

File: go/ql/test/library-tests/semmle/go/frameworks/Spew/TaintFlows.ql [hunk 2]
```diff
@@ -1,28 +1,18 @@
...
-class TestConfig extends TaintTracking::Configuration {
-  TestConfig() { this = "testconfig" }
-
-  override predicate isSource(DataFlow::Node source) { source instanceof UntrustedFlowSource }
+module TestConfig implements DataFlow::ConfigSig {
+  predicate isSource(DataFlow::Node source) {
+    source =
+      any(Function f | f.getName() = ["getUntrustedString", "getUntrustedStruct"])
+          .getACall()
+          .getResult()
```
Comment (mbg) [inline go/ql/test/library-tests/semmle/go/frameworks/Spew/TaintFlows.ql:9]: I note that the test previously ranged over all of `UntrustedFlowSource` but now is an inlined `UntrustedSource`. Do we miss out anything important by not having this range over all of `UntrustedFlowSource` anymore?

File: go/ql/test/library-tests/semmle/go/frameworks/Protobuf/TaintFlows.ql [hunk 3]
```diff
@@ -1,27 +1,18 @@
...
-
-class TestConfig extends TaintTracking::Configuration {
-  TestConfig() { this = "testconfig" }
-
-  override predicate isSource(DataFlow::Node source) { source instanceof UntrustedFlowSource }
+module Config implements DataFlow::ConfigSig {
+  predicate isSource(DataFlow::Node source) {
+    exists(Function fn | fn.hasQualifiedName(_, ["getUntrustedString", "getUntrustedBytes"]) |
+      source = fn.getACall().getResult()
+    )
```
Comment (mbg) [inline go/ql/test/library-tests/semmle/go/frameworks/Protobuf/TaintFlows.ql:8]: This previously ranged over all of `UntrustedFlowSource`, but is now more limited. Do we miss out on anything as a result of this change?

File: go/ql/test/library-tests/semmle/go/frameworks/K8sIoApiCoreV1/main.go [hunk 4]
```diff
@@ -9,86 +9,88 @@ import (
 //go:generate depstubber -vendor k8s.io/apimachinery/pkg/runtime ProtobufMarshaller,ProtobufReverseMarshaller
 
 func source() interface{} {
-	return make([]byte, 1, 1)
+	return make([]byte, 1)
```
Comment (mbg) [inline go/ql/test/library-tests/semmle/go/frameworks/K8sIoApiCoreV1/main.go:12]: What's the reason for this change?

File: go/ql/test/library-tests/semmle/go/frameworks/EvanphxJsonPatch/TaintFlows.ql [hunk 5]
```diff
@@ -1,33 +1,20 @@
...
-      value = "" and
-      sink.hasLocationInfo(location.getFile().getAbsolutePath(), location.getStartLine(),
-        location.getStartColumn(), location.getEndLine(), location.getEndColumn())
+  predicate isSink(DataFlow::Node sink) {
+    exists(Function fn | fn.hasQualifiedName(_, ["sinkByteArray", "sinkPatch"]) |
+      sink = fn.getACall().getAnArgument()
     )
   }
+
+  int fieldFlowBranchLimit() { result = 1000 }
```
Comment (mbg) [inline go/ql/test/library-tests/semmle/go/frameworks/EvanphxJsonPatch/TaintFlows.ql:17]: Could you add a comment noting why you added this here for this particular configuration?

**Review by mbg — COMMENTED** [review]
Good work updating all of this! ✨ The refactoring generally looks reasonable. There are a few specific comments I had while looking over this and two more broad ones:

- For queries where the expected test output changed and you haven't done so yet, it would be good to note why it changed in the commit description or in the PR description.
- Most/all of the old data/taint flow configurations had docs comments explaining their purpose. However, the new `ConfigSig` implementations and data/taint flow configurations mostly do not have matching docs comments. In general, it would be nice to have more documentation/comments.

There also seem to a good few `ConfigSig`s which are basically the same as a result of the `[..]Customizations` approach along the lines of:
```ql
private module Config implements DataFlow::ConfigSig {
    predicate isSource(DataFlow::Node source) { source instanceof Source }

    predicate isSink(DataFlow::Node sink) { sink instanceof Sink }

    predicate isBarrier(DataFlow::Node node) { node instanceof Barrier }
}
```
Would it be possible (in the future, not as part of this PR) to refactor this so that we have a generic implementation that we can just parameterise over the module that's providing the `Source`, `Sink`, and `Barrier`?

File: go/ql/src/Security/CWE-020/MissingRegexpAnchor.ql [hunk 6]
```diff
@@ -60,22 +60,22 @@ predicate isInterestingUnanchoredRegexpString(string re, string msg) {
       "hosts may come before or after it."
 }
 
-class Config extends DataFlow::Configuration {
-  Config() { this = "MissingRegexpAnchor::Config" }
-
-  predicate isSourceString(DataFlow::Node source, string msg) {
+module Config implements DataFlow::ConfigSig {
+  additional predicate isSourceString(DataFlow::Node source, string msg) {
```
[the remark this answers is not in the document]
Comment (owen-mc) [reply go/ql/src/Security/CWE-020/MissingRegexpAnchor.ql:64]: The only difference I'm aware of is that if you want to make the module private then you have to move the predicate out of it.

File: go/ql/test/library-tests/semmle/go/dataflow/InterProceduralDataFlow/Test.ql
[the remark this answers is not in the document]
Comment (owen-mc) [reply go/ql/test/library-tests/semmle/go/dataflow/InterProceduralDataFlow/Test.ql:1]: I wanted to make this use InlineFlowTest, or failing that InlineExpectationsTest, but I couldn't easily do it. I didn't want to throw away a lot of the information that is currently in the .expected file. But that's okay, it's okay to have tests which aren't InlineFlowTests or InlineExpectationsTests.

File: go/ql/lib/semmle/go/security/StringBreak.qll [hunk 7]
```diff
@@ -31,4 +33,21 @@ module StringBreak {
 
     override predicate isSanitizer(DataFlow::Node nd) { quote = nd.(Sanitizer).getQuote() }
   }
+
+  private module Config implements DataFlow::StateConfigSig {
+    /** The type of quote being tracked by this configuration. */
+    class FlowState = Quote;
+
+    predicate isSource(DataFlow::Node source, FlowState state) {
+      source instanceof Source and exists(state)
```
[the remark this answers is not in the document]
Comment (owen-mc) [reply go/ql/lib/semmle/go/security/StringBreak.qll:42]: Just to make the warning that `state` isn't used go away. Maybe it's okay to have that warning.

File: go/ql/test/library-tests/semmle/go/frameworks/K8sIoApiCoreV1/main.go (hunk 4 above)
Comment (owen-mc) [reply go/ql/test/library-tests/semmle/go/frameworks/K8sIoApiCoreV1/main.go:12]: My IDE told me it was better to do it this way. Presumably it's a built-in Go linter of some kind.

File: go/ql/test/query-tests/Security/CWE-190/AllocationSizeOverflow.expected
[the remark this answers is not in the document]
Comment (owen-mc) [reply go/ql/test/query-tests/Security/CWE-190/AllocationSizeOverflow.expected:1]: It generally isn't obvious why. I know one cause is when there were more than one config in scope in the old version. I'm not sure how much value there is in tracking it down. I'm going to bet that in the corresponding PRs for java and C# they have similar changes and they haven't looked into what caused them. (Of course if there was a change in the results it would be different. It's just because the only change is in path nodes and edges.)

File: config/identical-files.json [hunk 8]
```diff
@@ -603,4 +602,4 @@
     "python/ql/lib/semmle/python/security/internal/EncryptionKeySizes.qll",
     "java/ql/lib/semmle/code/java/security/internal/EncryptionKeySizes.qll"
   ]
-}
+}
```
[the remark this answers is not in the document]
Comment (owen-mc) [reply config/identical-files.json:605]: No

File: go/ql/test/library-tests/semmle/go/dataflow/VarArgsWithFunctionModels/Flows.ql [hunk 9]
```diff
@@ -18,69 +16,15 @@ class SummaryModelTest extends DataFlow::FunctionModel {
...
-    // (inp.isParameter(0) and outp.isResult)
-    // or
-    // this.hasQualifiedName("github.com/nonexistent/test", "FunctionWithVarArgsOfStructsParameter")) and
-    // (inp.isParameter(0) and outp.isResult)
+    or
+    this.hasQualifiedName("github.com/nonexistent/test", "FunctionWithSliceOfStructsParameter") and
+    (inp.isParameter(0) and outp.isResult())
+    or
+    this.hasQualifiedName("github.com/nonexistent/test", "FunctionWithVarArgsOfStructsParameter") and
+    (inp.isParameter(0) and outp.isResult())
```
[the remark this answers is not in the document]
Comment (owen-mc) [reply go/ql/test/library-tests/semmle/go/dataflow/VarArgsWithFunctionModels/Flows.ql:24]: I added this comment in the test
```
	// The following tests all fail because FunctionModel doesn't interact with access paths
```
I believe that when I originally wrote the tests and realised this I just left the models commented out. This time I thought it was better to have the models there and the tests still failing. I don't know what's best. Maybe just removing those tests, and having a comment noting this particular shortcoming of `FunctionModel`?

File: go/ql/src/Security/CWE-089/StringBreak.expected
[the remark this answers is not in the document]
Comment (owen-mc) [reply go/ql/src/Security/CWE-089/StringBreak.expected:1]: Oops, this is from running the query like it was a test. It analyzes the examples from the qhelp. I'll delete it.

File: go/ql/lib/semmle/go/security/UnsafeUnzipSymlink.qll [hunk 10]
```diff
@@ -33,18 +35,31 @@ module UnsafeUnzipSymlink {
...
+  // Archive header field symlinks resolved
+  private module EvalSymlinksConfig implements DataFlow::ConfigSig {
+    predicate isSource(DataFlow::Node source) { source instanceof FilenameWithSymlinks }
+
+    predicate isSink(DataFlow::Node sink) { sink instanceof EvalSymlinksSink }
+
+    predicate isBarrier(DataFlow::Node node) { node instanceof EvalSymlinksInvalidator }
+  }
+
+  private module EvalSymlinksFlow = TaintTracking::Global<EvalSymlinksConfig>;
```
[the remark this answers is not in the document]
Comment (owen-mc) [reply go/ql/lib/semmle/go/security/UnsafeUnzipSymlink.qll:47]: Are you saying that I should keep it public? I agree that that would be more backwards-compatible. But as this is a helper I think it should always have been private.

File: go/ql/src/experimental/CWE-74/DsnInjectionCustomizations.qll [hunk 11]
```diff
@@ -25,6 +28,20 @@ class DsnInjection extends TaintTracking::Configuration {
...
 }
 
+private module DsnInjectionConfig implements DataFlow::ConfigSig {
+  predicate isSource(DataFlow::Node source) { source instanceof Source }
+
+  predicate isSink(DataFlow::Node sink) {
+    exists(Function f | f.hasQualifiedName("database/sql", "Open") |
+      sink = f.getACall().getArgument(1)
+    )
+  }
```
[the remark this answers is not in the document]
Comment (owen-mc) [reply go/ql/src/experimental/CWE-74/DsnInjectionCustomizations.qll:38]: No. That line was added very recently, after I'd converted it, and I didn't spot it when I rebased. Well spotted!

File: go/ql/test/library-tests/semmle/go/frameworks/Spew/TaintFlows.ql (hunk 2 above)
Comment (owen-mc) [reply go/ql/test/library-tests/semmle/go/frameworks/Spew/TaintFlows.ql:9]: I don't think so - we're just trying to match the two functions used as sources in the test files. I can only imagine was written like this to be more like a real source definition.

File: go/ql/test/library-tests/semmle/go/frameworks/Protobuf/TaintFlows.ql (hunk 3 above)
Comment (owen-mc) [reply go/ql/test/library-tests/semmle/go/frameworks/Protobuf/TaintFlows.ql:8]: As above, I don't think it matters for tests.

File: go/ql/test/library-tests/semmle/go/frameworks/EvanphxJsonPatch/TaintFlows.ql (hunk 5 above)
Comment (owen-mc) [reply go/ql/test/library-tests/semmle/go/frameworks/EvanphxJsonPatch/TaintFlows.ql:17]: It isn't needed. I've removed it. I must have added it by mistake, probably copying from somewhere else.

File: go/ql/src/Security/CWE-020/MissingRegexpAnchor.ql (hunk 6 above)
[the remark this answers is not in the document]
Comment (mbg) [reply go/ql/src/Security/CWE-020/MissingRegexpAnchor.ql:64]: I suppose in any case it would be good to have consistency in which approach is used.

**Comment by owen-mc** [comment]
I believe I've addressed all comments except explaining the changes in test expectations.

**Comment by owen-mc** [comment]
I've added explanations of test result changes in the relevant commit messages. I had to force-push to do this, so I rebased on `main` at the same time. There were no changes to the commits except the commit messages.

File: go/ql/lib/semmle/go/security/LogInjection.qll [hunk 12]
```diff
@@ -15,6 +15,8 @@ module LogInjection {
   import LogInjectionCustomizations::LogInjection
 
   /**
+   * DEPRECATED: Use `Flow` instead.
+   *
```
Comment (mbg) [inline go/ql/lib/semmle/go/security/LogInjection.qll:19]: Can't comment on the relevant line because it wasn't changed as part of this PR, but `Flow` here doesn't have any documentation.

File: go/ql/lib/semmle/go/security/OpenUrlRedirect.qll [hunk 13]
```diff
@@ -63,4 +65,45 @@ module OpenUrlRedirect {
...
+      // block propagation of this unsafe value when its host is overwritten
+      exists(Write w, Field f | f.hasQualifiedName("net/url", "URL", "Host") |
+        w.writesField(node.getASuccessor(), f, _)
+      )
+      or
+      hostnameSanitizingPrefixEdge(node, _)
+    }
+  }
+
+  /** Tracks taint flow for reasoning about unvalidated URL redirections. */
```
Comment (mbg) [inline go/ql/lib/semmle/go/security/OpenUrlRedirect.qll:107]:
```suggestion
  /** Tracks taint flow from unvalidated, untrusted data to URL redirections. */
```

File: go/ql/lib/semmle/go/security/ReflectedXss.qll [hunk 14]
```diff
@@ -35,4 +37,15 @@ module ReflectedXss {
...
+
+  private module Config implements DataFlow::ConfigSig {
+    predicate isSource(DataFlow::Node source) { source instanceof Source }
+
+    predicate isSink(DataFlow::Node sink) { sink instanceof Sink }
+
+    predicate isBarrier(DataFlow::Node node) { node instanceof Sanitizer }
+  }
+
+  /** Tracks taint flow for reasoning about XSS. */
```
Comment (mbg) [inline go/ql/lib/semmle/go/security/ReflectedXss.qll:49]:
```suggestion
  /** Tracks taint flow from untrusted data to XSS attack vectors. */
```

File: go/ql/lib/semmle/go/security/RequestForgery.qll [hunk 15]
```diff
@@ -47,4 +49,24 @@ module RequestForgery {
...
+
+    predicate isAdditionalFlowStep(DataFlow::Node pred, DataFlow::Node succ) {
+      // propagate to a URL when its host is assigned to
+      exists(Write w, Field f, SsaWithFields v | f.hasQualifiedName("net/url", "URL", "Host") |
+        w.writesField(v.getAUse(), f, pred) and succ = v.getAUse()
+      )
+    }
+  }
+
+  /** Tracks taint flow for reasoning about request forgery. */
```
Comment (mbg) [inline go/ql/lib/semmle/go/security/RequestForgery.qll:70]:
```suggestion
  /** Tracks taint flow from untrusted data to request forgery attack vectors. */
```

File: go/ql/lib/semmle/go/security/StringBreak.qll (hunk 7 above)
[the remark this answers is not in the document]
Comment (mbg) [reply go/ql/lib/semmle/go/security/StringBreak.qll:42]: I have been trawling through the CodeQL docs and Slack a bit to see if there's a better way to do this, but couldn't find anything.

File: go/ql/lib/semmle/go/security/StringBreak.qll [hunk 16]
```diff
@@ -31,4 +33,25 @@ module StringBreak {
...
+    predicate isSink(DataFlow::Node sink, FlowState state) { state = sink.(Sink).getQuote() }
+
+    predicate isBarrier(DataFlow::Node node, FlowState state) {
+      state = node.(Sanitizer).getQuote()
+    }
+  }
+
+  /**
+   * Tracks taint flow for reasoning about unsafe-quoting vulnerabilities,
+   * parameterized with the type of quote being tracked.
```
Comment (mbg) [inline go/ql/lib/semmle/go/security/StringBreak.qll:54]:
```suggestion
   * Tracks taint flow from untrusted data which may contain single or double quotes to uses where those quotes need to be escaped. 
   * The type of quote is accessible through the `Sink`.
```

File: go/ql/lib/semmle/go/security/UnsafeUnzipSymlink.qll [hunk 17]
```diff
@@ -14,9 +14,11 @@ module UnsafeUnzipSymlink {
   import UnsafeUnzipSymlinkCustomizations::UnsafeUnzipSymlink
 
   /**
+   * DEPRECATED: Use a copy of `EvalSymlinksConfig` and `EvalSymlinksFlow` instead.
```
Comment (mbg) [inline go/ql/lib/semmle/go/security/UnsafeUnzipSymlink.qll:17]:
```suggestion
   * DEPRECATED: Use copies of `EvalSymlinksConfig` and `EvalSymlinksFlow` instead.
```

File: go/ql/lib/semmle/go/security/UnsafeUnzipSymlink.qll (hunk 10 above)
[the remark this answers is not in the document]
Comment (mbg) [reply go/ql/lib/semmle/go/security/UnsafeUnzipSymlink.qll:51]: It was the same comment as for the other file where the configuration was public and the `Flow` is private, so it is the same resolution here as well, which I think you have already done.

File: go/ql/src/experimental/CWE-1004/CookieWithoutHttpOnly.ql [hunk 18]
```diff
@@ -78,9 +78,9 @@ predicate isGorillaSessionsCookieFlow(
           sessionSave2.asPathNode2()) and
         (
           not BoolToGorillaSessionOptionsTrackingFlow::flowTo(sink.getNode()) and
-          source = options and
           sink = sessionSave2 and
-          sessionSave2.getNode() = sessionSave.getNode()
+          source = options and
+          sessionSave.getNode() = sessionSave2.getNode()
```
Comment (mbg) [inline go/ql/src/experimental/CWE-1004/CookieWithoutHttpOnly.ql:83]: Swap the two sides around for even fewer unnecessary changes!

**Review by mbg — APPROVED** [review]
Nice work investigating the reasons for the changes in the expected test output! Those comments are really useful, and more varied than I expected. 

I think this is probably good to merge as is. I have added a few minor bits of feedback for the new documentation, which might be nice to spend to improve if you have time, but otherwise aren't urgent changes.

Other comments:

- 855677c62514138f8eddafbfc0ded819b6204330 has meaningful changes to the `.expected` files, but no explanation. 
- b0302b6da09e1ec402b3effbeb5dd14a87dd9aa0 still has a stray `f` at the end of the commit message.

File: go/ql/lib/semmle/go/security/CleartextLogging.qll [hunk 19]
```diff
@@ -56,4 +58,43 @@ module CleartextLogging {
...
+      not any(Protobuf::GetMethod gm).taintStep(src, trg)
+    }
+  }
+
+  /**
+   * Tracks data flow for reasoning about clear-text logging of sensitive
+   * information, from `Source`s, which are sources of sensitive data, to
+   * `Sink`s, which is an abstract class representing all the places sensitive
+   * data may be stored in cleartext. Additional sources or sinks can be added
+   * by extending the relevant class.
```
[the remark this answers is not in the document]
Comment (owen-mc) [reply go/ql/lib/semmle/go/security/CleartextLogging.qll:97]: This is copied from elsewhere in the file. I'm sure there are many improvements we could make to our documentation, but I suggest this could be done in a separate PR afterwards.

**Review by mbg — APPROVED** [review]
🚢
